### PR TITLE
SI-7883 - don't iterate over all keys in MapWrapper.containsKey()

### PR DESCRIPTION
**Pull request by ivmaykov**
This commit fixes SI-7883. I've checked that the new test fails without the changes in Wrappers.scala and passes with them.

### CODE REVIEW

**Comment by adriaanm** [comment]
This would be faster/nicer as a junit test. Check out https://github.com/scala/scala/commit/834e29ff7f2047054e4af4c05311943bcbe480a7 for an example.

**Comment by adriaanm** [comment]
Also -- where are my manners -- thanks for the fix!

**Comment by ivmaykov** [comment]
@adriaanm - I can rewrite the test using JUnit, I just wasn't sure before where a test like this should go and saw a bunch of "prior art" in test/files/run. Which directory should new JUnit tests be added to?

**Comment by adriaanm** [comment]
There's an example in the commit I linked to above.

**Comment by adriaanm** [comment]
It's a pretty new way of doing things, but we'd like to junit whatever we
can junit. test/files/junit is the place to be

**Comment by ivmaykov** [comment]
Hmm, do you mean test/junit/scala/...? If so, should this go under test/junit/scala/collection? Would you prefer I call it Si7883Test or MapWrapperTest? If I go with MapWrapperTest, it would start out as very incomplete and could eventually contain more test cases for MapWrapper I suppose, but we lose the one-test-file-per-issue granularity. Please advise.

**Comment by adriaanm** [comment]
For unit tests I think a class granularity makes more sense than test
granularity.

**Comment by ivmaykov** [comment]
@Ichoran - Scala's Map.containsKey() method takes the key type (A), while Java's Map.contains() takes an Object (aka AnyRef in scala). This means that, for example, for a Map[String, _], the Scala method can only be called with a String argument, but the Java method can be called with any argument. If we call the java method with a non-String argument (say, an Int), we'd get the ClassCastException and return false, which makes sense - a Map with String keys cannot contain an Int key.

**Comment by Ichoran** [comment]
The Scala collection.Map signature is, in bytecode,
  public abstract boolean contains(java.lang.Object);
because it is generic.  There is no cast; the Int is already boxed as an object by that point.

**Comment by ivmaykov** [comment]
@adriaanm - ok, will do

**Comment by ivmaykov** [comment]
@Ichoran - as far as I can tell, the cast is needed to make this work. See the contains() method at scala.collection.MapLike.scala:150. When I remove the cast, the call ends up going to java.util.AbstractCollection.contains()

**Comment by Ichoran** [comment]
The cast doesn't actually do anything, though, except for tell the compiler to pretend that its java.lang.Object is a generic type instead of java.lang.Object.  There's no bytecode associated.  So there's no exception to throw.  (Can you come up with an example that demonstrates otherwise?)

File: test/files/run/si7883.scala [hunk 1]
```diff
@@ -0,0 +1,24 @@
+object Test extends App {
+  import scala.collection.JavaConverters.mapAsJavaMapConverter
+
+  // A HashMap which throws an exception when the iterator() method is called.
+  // Before the fix for SI-7883, calling MapWrapper.containsKey() used to
+  // iterate through every element of the wrapped Map, and thus would crash
+  // in this case.
+  val scalaMap = new scala.collection.mutable.HashMap[String, String] {
+    override def iterator = throw new UnsupportedOperationException
```
Comment (viktorklang) [inline test/files/run/si7883.scala]: Add a counter that is incremented each time contains is called and add a test to verify that it is called once per containsKey call?

Comment (ivmaykov) [inline test/files/run/si7883.scala]: Will do in updated pull request

**Comment by ivmaykov** [comment]
@adriaanm - I've converted the test to junit.

**Comment by ivmaykov** [comment]
@Ichoran - ah, so you mean that we can do the cast and not need the `try/catch`? I'll give it a shot ... I figured it was necessary as implied by MutableMapWrapper.remove() in the same file.

**Comment by ivmaykov** [comment]
@Ichoran - removing the try/catch in MapWrapper results in the following when we call `javaMap.containsKey(42)`:

```
Testsuite: scala.collection.convert.MapWrapperTest
Tests run: 1, Failures: 0, Errors: 1, Time elapsed: 0.088 sec

Testcase: testContains took 0.076 sec
        Caused an ERROR
java.lang.Integer cannot be cast to java.lang.String
java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.String
        at scala.collection.convert.MapWrapperTest$$anon$1.contains(MapWrapperTest.scala:21)
        at scala.collection.convert.Wrappers$MapWrapper.containsKey(Wrappers.scala:216)
        at scala.collection.convert.MapWrapperTest.testContains(MapWrapperTest.scala:41)
```

**Comment by ivmaykov** [comment]
Please review latest commit 55a558a

File: test/files/run/si7883.scala (hunk 1 above)
Comment (ivmaykov) [inline test/files/run/si7883.scala]: Done in 55a558a

**Comment by adriaanm** [comment]
you can run a single test by specifying the scala file to compile and run: `test/partest test/junit/scala/collection/convert/MapWrapperTest.scala`

**Comment by ivmaykov** [comment]
@adriaanm - that didn't work for junit tests for me for some reason (did work for the old test in `test/files/run`), but `ant test.junit` did the trick and was fast enough.

**Comment by Ichoran** [comment]
@ivmaykov - LGTM.  I forgot that subclassing to a specific type would allow you to replace the generic method with a specific one and then generate the cast exception.  Thanks for working that through!

**Comment by som-snytt** [comment]
When `partest` does unit tests, it will be `punit` as in `punitive`, but for now one must choose one's favorite version of ant 1.8 or IDE.

**Comment by ivmaykov** [comment]
Got the LGTM, but the build bot has not emailed me about the latest commit (55a558a). It did complain about an old commit failing (5222ecd), but that's been fixed in 55a558a. Is there a way to force the build bot to build the latest commit?

**Comment by ivmaykov** [comment]
PLS REBUILD/pr-scala@55a558a

**Comment by ivmaykov** [comment]
Build passed (https://scala-webapps.epfl.ch/jenkins/job/pr-scala/3633/) and have LGTM, feel free to merge anytime.

**Comment by adriaanm** [comment]
>  the build bot has not emailed me about the latest commit

it doesn't tell you about your successes, only your failures -- success is implicit in the green checkmark next to a commit

**Comment by adriaanm** [comment]
@ivmaykov, as it turned out to be tricky to understand why we need this `try/catch`, could you please follow up with a second commit that renders the discussion here as a code comment? Thanks, and congrats already on your first (I believe) contribution to Scala!

**Comment by Ichoran** [comment]
Something like "Subclass of HashMap with specific type may redirect generic
contains to specific contains, which will throw a ClassCastError if the
wrong type is passed" would be great.

**Comment by ivmaykov** [comment]
@adriaanm, @Ichoran - added comment, see f38f762

**Comment by ivmaykov** [comment]
@adriaanm - that might be sufficient, but I think as written the code is a bit cleaner - we are doing the cast and catching the `ClassCastException` in the same place. I believe you are suggesting this:

```
override def containsKey(key: AnyRef): Boolean = try {
  underlying.contains(key)
} catch {
  case ex: ClassCastException => false
}
```

Which separates the type cast from the catch and in my opinion reduces readability. Even if the cast is actually a no-op and the real cast happens inside `contains`, I think adding the cast as written looks better. The code as written also follows conventions established in the same file, for example in `MutableMapWrapper.remove`:

```
override def remove(k: AnyRef): B = try {
  underlying remove k.asInstanceOf[A] match {
    case None => null.asInstanceOf[B]
    case Some(v) => v
  }
} catch {
  case ex: ClassCastException => null.asInstanceOf[B]
}
```

I can remove the cast in `containsKey` if you're really against it, but I think the code becomes less clear that way.

**Comment by adriaanm** [comment]
In your second example, the cast is actually necessary for the code to type check. In your fix, it isn't. I'm not against having the catch as documentation, but then I'd prefer the comment to be clarified. Right now it implies the cast is necessary ("This is why we need a type cast to A inside a try/catch." -- we don't _need_ it, though)

**Comment by ivmaykov** [comment]
Actually if I remove the cast, then the wrong `contains` ends up being called and the test fails, check it out:

```
Testsuite: scala.collection.convert.MapWrapperTest
Tests run: 1, Failures: 0, Errors: 1, Time elapsed: 0.091 sec

Testcase: testContains took 0.078 sec
        Caused an ERROR
null
java.lang.UnsupportedOperationException
        at scala.collection.convert.MapWrapperTest$$anon$1.iterator(MapWrapperTest.scala:23)
        at scala.collection.convert.MapWrapperTest$$anon$1.iterator(MapWrapperTest.scala:21)
        at scala.collection.convert.Wrappers$IterableWrapperTrait$class.iterator(Wrappers.scala:25)
        at scala.collection.convert.Wrappers$IterableWrapper.iterator(Wrappers.scala:51)
        at scala.collection.convert.Wrappers$IterableWrapper.iterator(Wrappers.scala:51)
        at java.util.AbstractCollection.contains(AbstractCollection.java:82)
        at scala.collection.convert.Wrappers$MapWrapper.containsKey(Wrappers.scala:220)
        at scala.collection.convert.MapWrapperTest.testContains(MapWrapperTest.scala:37)
```

Edit: updated with a JDK 6 stack trace since that's recommended for scala development, previous trace was run under JDK 7.

**Comment by adriaanm** [comment]
wow... weird! sounds like we were right to dig a little deeper -- i'll also have a look to see what's going on...

**Comment by ivmaykov** [comment]
Yeah, I can't tell exactly how it's calling `j.u.AbstractCollection.contains`, my guess is through the `WrapAsJava._ / WrapAsScala._` implicit conversion imports at the top of the file. `MapWrapper` only extends `j.u.AbstractMap`, which shouldn't be a subclass of `AbstractCollection` according to http://docs.oracle.com/javase/6/docs/api/java/util/AbstractMap.html, so it shouldn't be able to call that method directly.

**Comment by adriaanm** [comment]
I've been playing around with this on both Java 6 and Java 7 and Scala 2.10.3 and 2.11.0-M5, but can't trigger the CCE. The code doesn't even compile for me:

```
scala>  class MapWrapper[A, B](underlying: Map[A, B]) extends java.util.AbstractMap[A, B] { self =>
     |     override def containsKey(key: AnyRef): Boolean =
     |       underlying.contains(key)
     |  }
<console>:10: error: type mismatch;
 found   : key.type (with underlying type AnyRef)
 required: A
             underlying.contains(key)
                                 ^
```

I've tried various approaches, but not sure what's going on here. I don't want to push this too far, as there's a lot of precedent for just casting our way out in these wrappers, but if you're so inclined, I'd love to understand exactly what's goign on.

**Comment by ivmaykov** [comment]
@adriaanm - what if you do this first?

```
import scala.collection.convert.WrapAsScala._
import scala.collection.convert.WrapAsJava._
```

**Comment by adriaanm** [comment]
@JamesIry, could you take a look?

**Comment by gkossakowski** [comment]
@JamesIry: ping

**Comment by gkossakowski** [comment]
Haven't heard from @JamesIry so rescheduling for M7.

**Comment by adriaanm** [comment]
@ivmaykov, sorry this slipped! I was on vacation last week and the ping does not seem to have reached @JamesIry. Thanks for your patience. We'll get this merged for the next milestone!

**Comment by ivmaykov** [comment]
@adriaanm sounds good

**Comment by adriaanm** [comment]
LGTM, thanks again, @ivmaykov!

**Pull request merged**